pkg/operator/ceph/object/objectstore.go

```diff
@@ -56,108 +62,288 @@ type idType struct {
+func deleteRealmAndPools(objContext *Context, spec cephv1.ObjectStoreSpec) error {
+	if spec.IsMultisite() {
+		// get list of endpoints not including the endpoint of the object-store for the zone
+		zoneEndpointsList, err := getZoneEndpoints(objContext, objContext.Endpoint)
+		if err != nil {
+			return err
+		}
 
-	err = deleteRealm(context)
-	if err != nil {
-		return errors.Wrap(err, "failed to delete realm")
-	}
+		realmArg := fmt.Sprintf("--rgw-realm=%s", objContext.Realm)
+		zoneGroupArg := fmt.Sprintf("--rgw-zonegroup=%s", objContext.ZoneGroup)
+		zoneEndpoints := strings.Join(zoneEndpointsList, ",")
+		endpointArg := fmt.Sprintf("--endpoints=%s", zoneEndpoints)
 
-	lastStore := false
-	if len(stores) == 1 && stores[0] == context.Name {
-		lastStore = true
-	}
+		zoneIsMaster, err := checkZoneIsMaster(objContext)
+		zoneGroupIsMaster := false
+		if zoneIsMaster {
+
+			_, err = RunAdminCommandNoMultisite(objContext, "zonegroup", "modify", realmArg, zoneGroupArg, endpointArg)
+			if err != nil {
+				return errors.Wrapf(err, "failed to remove object store %q endpoint from rgw zone group %q", objContext.Name, objContext.ZoneGroup)
+			}
+			logger.Debugf("endpoint %v was removed from zone group %v. the remaining endpoints in the zone group are %v", objContext.Endpoint, objContext.ZoneGroup, zoneEndpoints)
 
-	if !spec.PreservePoolsOnDelete {
-		err = deletePools(context, spec, lastStore)
+			// check if zone group is master only if zone is master for creating the system user
+			zoneGroupIsMaster, err = checkZoneGroupIsMaster(objContext)
+			if err != nil {
+				return errors.Wrapf(err, "failed to find out whether zone group %q in is the master zone group", objContext.ZoneGroup)
+			}
+		}
+
+		_, err = runAdminCommand(objContext, "zone", "modify", endpointArg)
 		if err != nil {
-			return errors.Wrap(err, "failed to delete object store pools")
+			return errors.Wrapf(err, "failed to remove object store %q endpoint from rgw zone %q", objContext.Name, spec.Zone.Name)
 		}
-	} else {
-		logger.Infof("PreservePoolsOnDelete is set in object store %s. Pools not deleted", context.Name)
-	}
-	return nil
-}
+		logger.Debugf("endpoint %v was removed from zone %v. the remaining endpoints in the zone are %v", objContext.Endpoint, objContext.Zone, zoneEndpoints)
 
-func setMultisite(context *Context, serviceIP string, spec cephv1.ObjectStoreSpec, realmName string, zoneGroupName string, zoneName string) error {
-	realmArg := fmt.Sprintf("--rgw-realm=%s", realmName)
-	zoneGroupArg := fmt.Sprintf("--rgw-zonegroup=%s", zoneGroupName)
-	zoneArg := fmt.Sprintf("--rgw-zone=%s", zoneName)
-	endpointArg := fmt.Sprintf("--endpoints=%s:%d", serviceIP, spec.Gateway.Port)
-	updatePeriod := false
+		if zoneIsMaster && zoneGroupIsMaster && zoneEndpoints == "" {
+			logger.Infof("WARNING: No other zone in realm %v can commit to the period or pull the realm until you create another object-store in zone %v", objContext.Realm, objContext.Zone)
```

Comment:
@travisn this is the warning I give out to the user if they delete an the last object store in the master zone and master zone group of a realm.